### ISSUE TITLE
Add DTD for external readers of paraview .vtu format

### ISSUE THREAD

**Issue by YetAnotherMinion** [SCOREC/core#6]
The vtu files generated by `writeVtkFiles` seems to be XML, would it make sense to include a DTD for these so that external parsers can load the meshes for human readable editing? My project is looking for an easy way to edit tags on meshes manually after visual inspection from ParaView, and the vtu format appears to offer really easy way to do this. Most of the code I am writing is equivalent to a DTD in a validating parser, so I though it would be nice this feature built into the `.pvtu` and `.vtu` files.

Importantly, ParaView seems to ignore any type of DTD, even if it is malformed and incorrect. So this would not impact any existing workflows

Below is an example simple DTD I whipped up for a single `.vtu` file, it needs a little more work for all of the cases in apfVtk.cc but it seems very possible based on inspection of the source.

```
<!DOCTYPE VTKFile [
	<!ELEMENT VTKFile (UnstructuredGrid?)>
	<!ATTLIST VTKFile
		type (UnstructuredGrid) #REQUIRED>
	<!ELEMENT UnstructuredGrid (Piece*)> <!-- can have more than one <Piece> -->
	<!ATTLIST Piece
		NumberOfPoints CDATA #IMPLIED
		NumberOfCells CDATA #IMPLIED>
	<!ELEMENT Piece ((Cells| Points|PointData|CellData)?, 
	(Points|Cells|PointData|CellData)?,
	(Points|Cells|PointData|CellData)?,
	(Points|Cells|PointData|CellData)?)> <!-- only one of each type in any order -->
	<!ELEMENT Points (DataArray)>
	<!ELEMENT PointData (DataArray*)>
    <!ELEMENT DataArray (#PCDATA)>
    <!ATTLIST DataArray 
    	type (Float64|Int32|Int64|UInt8) #REQUIRED
    	Name CDATA #IMPLIED
    	NumberOfComponents CDATA #IMPLIED
    	format (ascii) #REQUIRED>
    <!ELEMENT Cells (DataArray*)>
    <!ELEMENT CellData (DataArray*)> 
    ]>
```

**Comment by ibaned** [comment]
1) please give some background on yourself and your project if you open issues on our project,
we're just genuinely interested in who is using it and for what.

2) I think if anyone is going to issue such a DTD it should be Kitware, Inc.
They define the format, we just generate files that conform to it.
My understanding (from five minutes of Wikipedia reading) is that
a DTD can be provided externally, it does not have to be inside
the XML file with the content.
If thats the case, I recommend maintaining this DTD yourself and
making it a separate input to your tools, then contacting Kitware
to see if they are willing to coordinate on it.
Basically, it should not be the job of the code that writes a format
to issue the DTD, it should be done by those who define and read
the format.

**Comment by ibaned** [comment]
by the way, this is where we get the format information from:

http://www.vtk.org/wp-content/uploads/2015/04/file-formats.pdf

Also, I'm not entirely opposed to having us print this,
we would just need a bit of background as I said and
some way to test/validate the DTD from time to time.

**Comment by YetAnotherMinion** [comment]
1) This is that woefully unprepared undergrad from FEP. I have finally gotten my stuff in gear and am finishing the FEA solver from assign 4. I am creating an input file format for mesh and boundary conditions so I can abstract the dirty details from my analysis code. 

What I am trying to do is reassemble the mesh and its numberings from the `.vtk` file format. That way I can define boundary conditions by editing the mesh file by hand or using simple script from ParaView. This will enable me to easily define the boundary conditions as scalar or vector valued numberings with special names that my application can read. I can also dump the mesh at each stage and inspect it visually when I am debugging routines, then insert it back into the program seamlessly. I am not looking to do really complex boundary conditions, just constant loads across edges and faces, so I thought that using the same XML file might be an easy way to accomplish this.
Then the application can perform the integration and assign each node boundary conditions

I have no previous experience with any mesh editing tools, so if this can be done already in a free package that can be subsequently read by your codes by all means let me know. Right now I am programmatically generating the mesh, but I would like a way to specify the boundary conditions interactively for small meshes rather than write specific code for each test case.

This is the type of interaction I am considering to mirror the public API of apfVtk.cc
```C++
static void readPvtuFile(const char* filename, apf::Mesh* m);
static void readVtuFile(const char* filename, apf::Numbering* n);
```

2)First off, thanks for the link to the format info, I was looking for a comprehensive description of the file format, but all I found was people conflating the format with VTK files and the visualization toolkit.

You are right about an external DTD being better, that is what I am using right now, but I have to inject something like the following `<!DOCTYPE VTKFile SYSTEM "VTKFile.dtd"> ` into beginning of each file when I read them and that just seemed hacky to me. I will look for a way to have the DTD definition injected by my parser in a better way. I saw the DTD as an easy way to eliminate all of the validation code that I was writing to parse the `.vtu` files. I am using xerces-c++ 3.1 with the default SAX2 validating parser.

**Comment by ibaned** [comment]
1) So, editing things in VTK file form is not a terrible idea, but it also may not be the best idea. The usual approach would be to use classification to establish boundary conditions. On the mesh generation side, our usual go-to tool is proprietary, but we can also read files from Gmsh: http://geuz.org/gmsh/, in case you get tired of meshing by hand. Finally, regarding your proposed functions, you should note that a PVTU file is just a header for several VTU files grouped together to describe a parallel partitioned mesh. Our structures are quite rich, and you may spend more time coding a VTK reader than it would take to code your boundary conditions directly on the data structure

2) if you are having to edit the XML files anyway, then that is more reason for us to help out a bit. If you look at that format document and revise your DTD based on it, I can throw in some code that will add it to our files.

**Comment by YetAnotherMinion** [comment]
1) A full mesh compatibility with the full format at first would be complex indeed. My plan was to use a couple interfaces to abstract each stage. The global dispatcher reads the `.pvtu` file and sends all pieces to other piece loader processes. In single case it just sends it to itself. Then each sub loader process loads each piece file into the mesh. I only implement the functionality I need right now, then later can fill in more to create a complete feature if wanted.

`gmi_tag` seemed like one candidate for attaching data to the geometric model entities, but I still need to correctly create the model entities from the mesh. I had tried to derive the model from the mesh previously, 

```C++
gmi_register_null();
  gmi_model* g = gmi_load(".null");
  apf::Mesh2* m = apf::makeEmptyMdsMesh(g, 2, false);

// generate the mesh here

  apf::deriveMdsModel(m);//this makes CAD model for classification
  //accept the changes
  m->acceptChanges();
  m->verify();

// then after this cannot find a vertex classified on a model vertex

apf::MeshIterator* it = m->begin(2);
  apf::MeshEntity* e;
  apf::Downward down;
  int numVertices;
  apf::Vector3 point;
  int elm_num;
  while((e = m->iterate(it))) {
    /*get the vertices on this face element*/
    numVertices = m->getDownward(e, 0, down);
    elm_num = apf::getNumber(num_elms, e, 0, 0);
    std::cout << elm_num << ": ";
    for(int index = 0; index < numVertices; index++) {
      int vertex_id = apf::getNumber(numbers, down[index],0 ,0);
      apf::ModelEntity* model_ent = m->toModel(down[index]);
      int dim = m->getModelType(model_ent);
      std::cout << " dim: " << dim << ' ';
      }
    std::cout << std::endl;

// the above loop finds correct model edge and face classifications, but never any model vertices?
```


but I was not finding any vertex mesh entities classified upon vertex model entities. I looked at GMI but there was less documentation than there was for APF and I did not find anything obvious. I can classify on edge and face model entities correctly, which covers 90% of the way, but I still need some way of adding Dirichlet boundary conditions to model vertices, so I can fix one corner for example.

**Comment by ibaned** [comment]
This is because deriveMdsModel is not capable of finding sharp
corners and edges (such a decision would be arbitrary and error-prone
for general geometries), so it only creates two CAD entities,
one for the boundary and one for the interior.
For almost any problem of interest, you need more detail than that.
The simplest thing to do is to use the null model and specify the
CAD entity for every mesh entity you make.
mesh->findModelEntity will create a CAD entity if it doesn't exist,
when using the null model.
